build-tools/src/main/java/org/elasticsearch/gradle/testclusters/apmserver/ApmServerBuildService.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.testclusters.apmserver;
+
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.OperationCompletionListener;
+
+import java.io.IOException;
+
+public abstract class ApmServerBuildService
+    implements
+        BuildService<BuildServiceParameters.None>,
+        OperationCompletionListener,
+        AutoCloseable {
+    private final Logger logger = Logging.getLogger(getClass());
+
+    private MockApmServer mockServer;
+
+    public ApmServerBuildService() {
+        logger.info("starting up apm server");
+        mockServer = new MockApmServer();
+        try {
+            mockServer.start();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            logger.info("shutting down apm server");
+            mockServer.stop();
+        }));
+    }
+
+    @Override
+    public void close() throws Exception {
+        logger.info("shutting down apm server");
+        mockServer.stop();
+    }
+
+    @Override
+    public void onFinish(FinishEvent finishEvent) {
+        logger.info("shutting down apm server");
```

Comment:
Is this actually getting called? I don't see where we are actually registering this listener.

Reply:
this is actually something I was trying to find out with @breskeby
I wanted to make sure that server.stop is called when the RunTask is finished.
Perhaps because the RunTask is simply 'killed' with sigterm it will never really finish and the ApmServerBuildService cannot call its close methods

note, that the stop is not really doing anything useful at the moment. This feature is for local development only and we might as well kill it sigterm and not do any cleanup (that is what is happening now)

Reply:
The run task will be killed because it is in the foreground (the "running" task in gradle). But the mock APM server is it's own process. How do you ensure that process is also shut down upon the server/gradle stopping?

Reply:
in this PR athe APMServer is implemented as a `BuildService<BuildServiceParameters.None>` which is then `tied` to `RunTask`as it is declaring the usage of it. 
gradle is meant to take care of this service lifecycle https://docs.gradle.org/current/userguide/build_services.html
I was expecting that it would be calling close/finished methods but I could not see this being invoked.

empirically  I can see that when I ctrl^c the run task the build service is also shut down.

I don't think the apmserver is a separate process in this setup (I did not see a separate process or anything actually related to ApmServerBuildService in the output of `ps`)
https://gist.github.com/pgomulka/97978cf72336b388473602601ea79566

Reply:
> empirically I can see that when I ctrl^c the run task the build service is also shut down.

I'm not sure this is behavior we can rely on. I don't see anything that is explicitly calling stop on that build service. There is nothing liking the `run` task to that build service. The only thing ^C is doing is cancelling the current _build_ which might also incidentally kill that server thread but that's likely implicit behavior. When you say the service is "shut down" did you confirm that `stop()` is being called?

Reply:
I can confirm that is **not** called from neither shutdownhook, onFinish (OperationCompletionListener)  or
close from AutoClosable

any ideas how to make it work?

Reply:
I think we need to implement what I linked above, which is to register that service as a listener (which we currently aren't doing). Just having it implement the listener interface isn't sufficient. We then need to interrogate the finish event to ensure it's the `run` task that's completing.

Reply:
@pgomulka here's an example of registering this service to the event listener registry: https://docs.gradle.org/current/userguide/build_services.html#other_ways_of_using_a_build_service  

Though I don't think this will work in how the run task is implemented as gradle doesn't handle ctrl+c to abort the task from running in a meaningful way. We probably could rework the run task to make that work with gradle continuous build support that would allow better lifecycle management

Reply:
I tried that with @mark-vieira yesterday and adding `    public abstract BuildEventsListenerRegistry getEventsListenerRegistry(); ` + `getEventsListenerRegistry().onTaskCompletion(serviceProvider);` to RunTask did not work

we wanted to add this to elasticsearch.run.gradle but it could not find getEventsListenerRegistry method

Reply:
After chatting with @breskeby it seems the problem here is that the `:run` task is not a typical task. It blocks, and the only way it completes is via a ^C which cancels the existing build. Doing so means that no task complete or even build complete events are fired. So it doesn't make sense to author this in a way that assumes you'll ever get one of those events. I think the emperical observation @pgomulka made is going to be our best bet. The run task is run in a thread in the daemon, and when that thread is terminated, so are any child threads, including the APM mock server in this case. Here's what I'd suggest:

We remove all the "build service" stuff. It's buying us nothing and introducing a bunch of complexity. We just directly spin up a mock server in the run task impl if the `--with-apm-server` is passed. We rely on the Gradle behavior of terminating threads properly when doing ^C.

@breskeby will investigate a more elegant and robust solution to the `:run` task at some point in the future.